architecture/2025-06-upgrades-gitops.md

```diff
@@ -0,0 +1,127 @@
+- Native Helm integration - failure automatically blocks upgrade
+- Works with all GitOps tools (Flux, ArgoCD, etc.) as well as Helm directly
+- Runs once per upgrade (not per pod replica)
+
+**Cons**:
+- Helm-specific, possibility of Helm breaking changes in the future
+- Have to build and integrate new container, `preflight`, into the Radius Helm chart
+- Code duplication between `preflight` container and `rad` CLI
+
+### Option 2: Init Container per Pod
+
+**How**: Add init container to each control plane pod running preflight checks.
+
+**Context**: [Init containers](https://kubernetes.io/docs/concepts/workloads/pods/init-containers/) run before app containers start. Each pod replica runs its own init container independently.
+
+**Real-world example**: [MySQL Helm chart](https://github.com/helm/charts/blob/master/stable/mysql/templates/deployment.yaml) uses init containers for data directory cleanup before MySQL starts.
+
+**Implementation**: Add to existing deployments in `/deploy/Chart/templates/*/deployment.yaml`:
+```yaml
+spec:
+  template:
+    spec:
+      initContainers:
+      - name: preflight
+        image: ghcr.io/radius-project/preflight:{{ .Chart.AppVersion }}
+        command: ["/preflight"]
+```
+
+**Pros**:
+- Reuses existing preflight code
+- Kubernetes-native, decouples from Helm
+
+**Cons**:
+- Runs redundantly on every pod replica
+- Can't prevent partial upgrades
+
+### Option 3: Upgrade Controller
+
+**How**: Custom controller watches GitOps resources, runs preflight checks, and coordinates upgrades.
+
+**Context**: Extends existing Radius controller (`pkg/controller/`) using [controller-runtime](https://pkg.go.dev/sigs.k8s.io/controller-runtime). Watches for GitOps resource changes and coordinates upgrade flow.
+
+**Implementation**: 
+```go
+// Flux support - watch HelmRelease resources
+func (r *FluxUpgradeReconciler) Reconcile(ctx context.Context, req reconcile.Request) {
+    // Watch HelmRelease.spec.chart.version changes
+    // Run preflight.Registry.RunAll() 
+    // Update HelmRelease.status.conditions to block/allow upgrade
+}
+
+// ArgoCD support - watch Application resources  
+func (r *ArgoUpgradeReconciler) Reconcile(ctx context.Context, req reconcile.Request) {
+    // Watch Application.spec.source.targetRevision changes
+    // Run preflight checks
+    // Update Application.status.conditions
+}
+```
+
+Add controller registrations in `cmd/controller/main.go`:
+```go
+// Requires separate reconcilers for each GitOps tool
+ctrl.NewControllerManagedBy(mgr).
+    For(&helmv2beta1.HelmRelease{}).   // Flux
+    Complete(&FluxUpgradeReconciler{})
+
+ctrl.NewControllerManagedBy(mgr).
+    For(&argov1alpha1.Application{}). // ArgoCD  
+    Complete(&ArgoUpgradeReconciler{})
+```
+
+**Pros**:
+- Simplest to implement and maintain, taps into Flux and ArgoCD resources and doesn't require large Radius Helm chart changes
+
+**Cons**:
+- Requires separate reconcilers for Flux vs ArgoCD
```

Comment:
Can we at least have a Base reconciler for different GitOps implementations? Like `base.go` and then `flux.go` and `argocd.go` may have different functions based on the need.